File: go/app/main/domain/mutual_attention_system.go
package domain

type MututalAttentionSystem struct {
    Users []*User
}

func NewMutualAttentionSysyem() *MututalAttentionSystem {
    return &MututalAttentionSystem{}
}

func (system *MututalAttentionSystem) AddUserByUsername(username string) {
    for _, _user := range system.Users {
        if _user.Username == username {
            return
        }
    }
    system.Users = append(system.Users, NewUser(username))
}

func (system *MututalAttentionSystem) AddUser(user *User) {
    for _, _user := range system.Users {
        if _user.ID == user.ID {
            return
        }
    }
    system.Users = append(system.Users, user)
}

func (system *MututalAttentionSystem) AddUserAll(users []*User) {
    for _, user := range users {
        system.AddUser(user)
    }
}

func (system *MututalAttentionSystem) GetUser(userId string) *User {
    for _, user := range system.Users {
        if user.ID.String() == userId {
            return user
        }
    }
    panic("Not exist user")
}

File: java/src/main/java/org/app/domain/MutualAttentionSystem.java
package org.app.domain;

import org.app.domain.exceptions.DuplicatedUserException;
import org.app.domain.exceptions.NotExistUserException;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class MutualAttentionSystem {
    private List<User> users;

    public MutualAttentionSystem() {
        this.users = new ArrayList<>();
    }

    public List<User> getUsers() {
        return users;
    }

    public void setUsers(List<User> users) {
        this.users = users;
    }

    public void addUser(String username){
        if (users.stream().anyMatch(user -> user.getUsername().equals(username))){
            throw new DuplicatedUserException();
        }
        users.add(new User(username));
    }

    public void addUserAll(List<User> users){
        List<User> newUsers = users.stream().filter(user -> !this.users.contains(user)).collect(Collectors.toList());
        this.users.addAll(newUsers);
    }

    public User getUser(String userId){
        Optional<User> user = users.stream().filter(tempUser -> tempUser.getId().toString().equals(userId)).findFirst();
        if (user.isEmpty()){
            throw new NotExistUserException();
        }
        return user.get();
    }
}
